### PR TITLE
Don't proceed with pod-state-checking steps immediately.

### PR DESCRIPTION
**Pull request by mml**
Wait until we're sure the list of "all the pods" is the right length,
then we can start waiting for them all to be ready.

This is a WIP, as I think my magical constant is bogus, and the test has other flaws related to this.  But I want to see what happens on jenkins.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/22831)

### CODE REVIEW

**Comment by mml** [comment]
@ixdy is there a way to force jenkins to run the Feature:Reboot and/or disruptive tests on this PR?  The test I'm fixing is both.

**Comment by ixdy** [comment]
@mml sadly, no, not right now.

**Comment by mml** [comment]
@ixdy OK, this is ready to review.

File: test/e2e/restart.go [hunk 1]
```diff
@@ -68,6 +68,9 @@ var _ = Describe("Restart [Disruptive]", func() {
 
 	It("should restart all nodes and ensure all nodes and pods recover", func() {
 		nn := testContext.CloudConfig.NumNodes
+		// Sadly, this count must be kept up to date.  This total is 13 system pods
```
Comment (ixdy) [inline test/e2e/restart.go]: :disappointed: is there no better way to probe for this?

Comment (ixdy) [inline test/e2e/restart.go]: at the very least this should probably be a constant, rather than tucked away inside a test case.

**Comment by ixdy** [comment]
fixes #22489 

cc @dchen1107 @bprashanth from the issue

**Comment by bprashanth** [comment]
There might be a better way to do this, you can ask the Kubelet for how many static pods it has, and wait till they all show up. We already have some testing endpoints, just add another and hit it via apiserver proxy (existing e2es do this for runningpods): 
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/server/server.go#L360

@kubernetes/goog-node might have opinions.

**Comment by yujuhong** [comment]
/runningpod shows you running containers grouped by pod UID, with limited information (e.g., image). There is no indication that this is a static pod, and you can still be surprised if kubelet hasn't started the containers yet when you check. /pods will give you a list of pods kubelet is syncing to (including static pods), and they'll have annotation "kubernetes.io/config.source: file".

The reboot tests (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/reboot.go) uses similar logic and we should update both tests to use the same functions.

**Comment by mml** [comment]
@yujuhong do you have an example of us querying /pods somewhere else?

**Comment by yujuhong** [comment]
The utility function of getting /pods is https://github.com/kubernetes/kubernetes/blob/482efba8def114f7bb815c19557f2f46b61ecf49/test/e2e/util.go#L3670
It's used in a couple of places.

**Comment by mml** [comment]
Awesome.  Thanks.

**Comment by ixdy** [comment]
@mml assigning back to you until this is ready for another pass

**Pull request closed without merging**